### ISSUE TITLE
Could not commit transaction due to exception during persistence

### ISSUE THREAD

**Issue by hariharshankar** [thinkaurelius/titan#271]
This exception is appearing again when I try to write to titan from multiple concurrent processes. I am using titan 0.3.1-SNAPSHOT 

``` text
Exception in thread "main" com.thinkaurelius.titan.core.TitanException: Could not commit transaction due to exception during persistence
        at com.thinkaurelius.titan.graphdb.transaction.StandardTitanTx.commit(StandardTitanTx.java:848)
        at com.thinkaurelius.titan.graphdb.blueprints.TitanBlueprintsGraph.commit(TitanBlueprintsGraph.java:41)
        at com.tinkerpop.blueprints.TransactionalGraph$commit.call(Unknown Source)
        at gov.lanl.egosystem.discovery.utils.BatchMSAcademic$_main_closure2.doCall(BatchMSAcademic.groovy:37)
        at sun.reflect.GeneratedMethodAccessor16.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:601)
        at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:90)
        at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:233)
        at org.codehaus.groovy.runtime.metaclass.ClosureMetaClass.invokeMethod(ClosureMetaClass.java:272)
        at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:903)
        at groovy.lang.Closure.call(Closure.java:415)
        at groovy.lang.Closure.call(Closure.java:428)
        at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:1379)
        at org.codehaus.groovy.runtime.DefaultGroovyMethods.each(DefaultGroovyMethods.java:1351)
        at org.codehaus.groovy.runtime.dgm$170.invoke(Unknown Source)
        at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite$PojoMetaMethodSiteNoUnwrapNoCoerce.invoke(PojoMetaMethodSite.java:271)
        at org.codehaus.groovy.runtime.callsite.PojoMetaMethodSite.call(PojoMetaMethodSite.java:53)
        at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:42)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:108)
        at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:116)
        at gov.lanl.egosystem.discovery.utils.BatchMSAcademic.main(BatchMSAcademic.groovy:32)
Caused by: com.thinkaurelius.titan.core.TitanException: Permanent exception during backend operation
        at com.thinkaurelius.titan.diskstorage.util.BackendOperation.execute(BackendOperation.java:64)
        at com.thinkaurelius.titan.graphdb.database.StandardTitanGraph.save(StandardTitanGraph.java:277)
        at com.thinkaurelius.titan.graphdb.transaction.StandardTitanTx.commit(StandardTitanTx.java:839)
        ... 21 more
Caused by: com.thinkaurelius.titan.diskstorage.locking.PermanentLockingException: Updated state: lock acquired but value has changed since read [LockClaim [backer=com.thinkaurelius.titan.diskstorage.locking.consistentkey.ConsistentKeyLockStore@787935a1, key=0x0000000000028c78, col=0x018a, expectedValue=0x014d6174657269616c7320536369656e63e55f3cc5]]
        at com.thinkaurelius.titan.diskstorage.locking.consistentkey.ConsistentKeyLockTransaction.verifyAllLockClaims(ConsistentKeyLockTransaction.java:330)
        at com.thinkaurelius.titan.diskstorage.locking.consistentkey.ConsistentKeyLockStore.mutate(ConsistentKeyLockStore.java:94)
        at com.thinkaurelius.titan.diskstorage.BackendTransaction.mutateVertexIndex(BackendTransaction.java:108)
        at com.thinkaurelius.titan.graphdb.database.IndexSerializer.addProperty(IndexSerializer.java:78)
        at com.thinkaurelius.titan.graphdb.database.StandardTitanGraph.persist(StandardTitanGraph.java:306)
        at com.thinkaurelius.titan.graphdb.database.StandardTitanGraph.access$000(StandardTitanGraph.java:45)
        at com.thinkaurelius.titan.graphdb.database.StandardTitanGraph$2.call(StandardTitanGraph.java:270)
        at com.thinkaurelius.titan.graphdb.database.StandardTitanGraph$2.call(StandardTitanGraph.java:203)
        at com.thinkaurelius.titan.diskstorage.util.BackendOperation.execute(BackendOperation.java:61)
```

**Comment by mbroecheler** [comment]
This is expected behavior. I have updated the Gotcha page to explain what is going on:
https://github.com/thinkaurelius/titan/wiki/Titan-Limitations

---

When defining unique Titan types with locking enabled (i.e. requesting that Titan ensures uniqueness) it is likely to encounter locking exceptions under concurrent modifications to the graph.

com.thinkaurelius.titan.diskstorage.locking.PermanentLockingException: Updated state: lock acquired but value has changed since read

Such exceptions are to be expected, since Titan cannot know how to recover from a transactional state where an earlier read value has been modified by another transaction since this may invalidate the state of the transaction. It most cases it is sufficient to simply re-run the transaction. If locking exceptions are very frequent, try to analyze and remove the source of congestion.
